src/object/clone.ts

```diff
@@ -26,48 +26,103 @@
-    return new Date(obj.getTime()) as T;
+    return new Date(obj.getTime()) as Shallowed<T>;
   }
 
   if (obj instanceof RegExp) {
-    return new RegExp(obj.source, obj.flags) as T;
+    return new RegExp(obj.source, obj.flags) as Shallowed<T>;
   }
 
   if (obj instanceof Map) {
     const result = new Map();
     for (const [key, value] of obj) {
       result.set(key, value);
     }
-    return result as T;
+    return result as Shallowed<T>;
   }
 
   if (obj instanceof Set) {
     const result = new Set();
     for (const value of obj) {
       result.add(value);
     }
-    return result as T;
+    return result as Shallowed<T>;
   }
 
-  if (typeof obj === "object") {
-    return Object.assign({}, obj) as T;
+  // BINARY DATA
+  if (obj instanceof Uint8Array) return new Uint8Array(obj) as Shallowed<T>;
+  if (obj instanceof Uint8ClampedArray) return new Uint8ClampedArray(obj) as Shallowed<T>;
+  if (obj instanceof Uint16Array) return new Uint16Array(obj) as Shallowed<T>;
+  if (obj instanceof Uint32Array) return new Uint32Array(obj) as Shallowed<T>;
+  if (obj instanceof BigUint64Array) return new BigUint64Array(obj) as Shallowed<T>;
+  if (obj instanceof Int8Array) return new Int8Array(obj) as Shallowed<T>;
+  if (obj instanceof Int16Array) return new Int16Array(obj) as Shallowed<T>;
+  if (obj instanceof Int32Array) return new Int32Array(obj) as Shallowed<T>;
+  if (obj instanceof BigInt64Array) return new BigInt64Array(obj) as Shallowed<T>;
+  if (obj instanceof Float32Array) return new Float32Array(obj) as Shallowed<T>;
+  if (obj instanceof Float64Array) return new Float64Array(obj) as Shallowed<T>;
+  if (obj instanceof ArrayBuffer) return obj.slice(0) as Shallowed<T>;
+  if (obj instanceof SharedArrayBuffer) return obj.slice(0) as Shallowed<T>;
+  if (obj instanceof DataView) return new DataView(obj.buffer.slice(0)) as Shallowed<T>;
+  if (obj instanceof Blob) return new Blob([obj], { type: obj.type }) as Shallowed<T>;
+  if (obj instanceof File) return new File([obj], obj.name, { type: obj.type }) as Shallowed<T>;
+
+  if (typeof obj === 'object') {
+    return Object.assign({}, obj) as Shallowed<T>;
   }
-  return obj;
+  return obj as Shallowed<T>;
 }
 
+export type Shallowed<T> = Equal<T, ShallowMain<T>> extends true ? T : ShallowMain<T>;
+type ShallowMain<T> = T extends [never]
+  ? never
+  : T extends object
+    ? T extends
+        | Array<any>
+        | Set<any>
+        | Map<any, any>
+        | Date
+        | RegExp
+        | Date
+        | Uint8Array
+        | Uint8ClampedArray
+        | Uint16Array
+        | Uint32Array
+        | BigUint64Array
+        | Int8Array
+        | Int16Array
+        | Int32Array
+        | BigInt64Array
+        | Float32Array
+        | Float64Array
+        | ArrayBuffer
+        | SharedArrayBuffer
+        | DataView
+        | Blob
+        | File
+      ? T
+      : OmitNever<{
+          [P in keyof T]: T[P] extends Function ? never : T[P];
+        }>
```

Comment:
I'm not sure what this type does here. Could you explain what this is doing? 

I thought cloning some objects with methods, like `clone({ foo: () => {} })` will also work. However this type definition does not seem to respect it.

Reply:
<img width="616" alt="image" src="https://github.com/user-attachments/assets/3c6a9613-6e91-42c8-bd20-e9b9a389efb5">

Reply:
There could be many edge cases, so we might just return `T` from `clone`. After all, it's well-known that shallow cloning class instances doesn't copy methods.

Reply:
A developer working on the code might know if a value has been shallow copied, but other collaborators may not be aware without inspecting the code (specifically, the function returning the cloned value). Therefore, I believe the type proposed by samchon can help reduce misunderstandings among developers.

Using this type enhances code readability and prevents unnecessary bugs and confusion within the team. By employing the Resolved type, we can avoid the misconception of certain methods existing on plain objects, ensuring that developers understand the correct nature of the cloned value and can use the code more accurately.

Reply:
<img width="739" alt="스크린샷 2024-07-17 오전 9 01 34" src="https://github.com/user-attachments/assets/56f356e9-5b33-4b85-833f-ee66c7896e39">
Additionally, as shown in the screenshot, lodash and structuredClone do not support function cloning. It would be clearer for the clone function in es-toolkit to also not support functions

Reply:
The `Shallowed<T>` type is required as @de-novo mentioned. If the source object type does not have any function property type, it `clone()` function will return the same type of original. Otherwise the source object some methods, the `Shallowed<T>` will remove the method properties.

By the way, if there's an object that defining functional property, I think it would better to change the logic of `clone()` function about the object type like below:

```typescript
return Object.fromEntries(
  Object.entries(obj).filter(([_key, value]) => value !== undefined && typeof value !== "function")
)
```

Reply:
@raon0211 In the typ level, functional property explicitly removed, otherwise in the variable level, it still alive as a `never` type. It does not occur compiler level bug because of the characteristics of the `never` type, but seems ugly. @kakasoo We need to study about it.

<img width="386" alt="image" src="https://github.com/user-attachments/assets/2f66dfd4-2763-449c-9f8f-fdeadab02894">

<img width="395" alt="image" src="https://github.com/user-attachments/assets/a6e7578e-d91c-4554-b622-687e64ede6ad">

Reply:
What will happen to `clone({ foo: () => {} })`? Here, as far as I know, we can correctly clone the `foo` property.

Reply:
> I understand that `clone` does not clone class instance's methods.
> 
> I was wondering what will happen to `clone({ foo: () => {} })`; Here, we can correctly clone the `foo` property. However it will be prohibited by types.
> 
> As far as I know, we cannot easily distinguish class methods and function properties, since TypeScript uses structural typing.

To fix the problem, changed the implementation code.

```typescript
return Object.fromEntries(
  Object.entries(obj).filter(([_key, value]) => value !== undefined && typeof value !== "function")
)
```

Reply:
I think changing the implementation is awkward since most people would expect that `foo` will be properly cloned. `lodash` will clone those properties, and in JavaScript function properties are cloned when shallow copying an object. 

```tsx
const object = { foo: () => {} )
const cloned = { ...object };
```

Reply:
In my opinion, unless there is a TypeScript type that simply handles this case, we allow this behavior, because I think cloning a class instance is not a major use case. 

Please refer to our [contributing guide](https://github.com/toss/es-toolkit/blob/main/.github/CONTRIBUTING.md#12-simplicity) where we value simple implementation, and our functions will have complex implementation to suit every usecase.